components/sup/src/templating/context.rs

```diff
@@ -95,81 +112,805 @@ impl<'a> RenderContext<'a> {
+            gossip_ip: sys.gossip_ip,
+            gossip_port: sys.gossip_port,
+            http_gateway_ip: sys.http_gateway_ip,
+            http_gateway_port: sys.http_gateway_port,
+            permanent: sys.permanent,
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////
+
+/// Templating proxy fro a `manager::service::Pkg` struct.
+///
+/// Currently exposed to users under the `pkg` key.
+#[derive(Clone, Debug)]
+struct Package<'a> {
+    ident: Cow<'a, PackageIdent>,
+    origin: Cow<'a, String>,
+    name: Cow<'a, String>,
+    version: Cow<'a, String>,
+    release: Cow<'a, String>,
+    deps: Cow<'a, Vec<PackageIdent>>,
+    env: Cow<'a, Env>,
+    // TODO (CM): Ideally, this would be Vec<u16>, since they're ports.
+    exposes: Cow<'a, Vec<String>>,
+    exports: Cow<'a, HashMap<String, String>>,
+    // TODO (CM): Maybe Path instead of Cow<'a PathBuf>?
+    path: Cow<'a, PathBuf>,
+    svc_path: Cow<'a, PathBuf>,
+    svc_config_path: Cow<'a, PathBuf>,
+    svc_data_path: Cow<'a, PathBuf>,
+    svc_files_path: Cow<'a, PathBuf>,
+    svc_static_path: Cow<'a, PathBuf>,
+    svc_var_path: Cow<'a, PathBuf>,
+    svc_pid_file: Cow<'a, PathBuf>,
+    svc_run: Cow<'a, PathBuf>,
+    svc_user: Cow<'a, String>,
+    svc_group: Cow<'a, String>,
+}
+
+impl<'a> Package<'a> {
+    fn from_pkg(pkg: &'a Pkg) -> Self {
+        Package {
+            ident: Cow::Borrowed(&pkg.ident),
+            // TODO (CM): have Pkg use FullyQualifiedPackageIdent, and
+            // get origin, name, version, and release from it, rather
+            // than storing each individually; I suspect that was just
+            // for templating
+            origin: Cow::Borrowed(&pkg.origin),
+            name: Cow::Borrowed(&pkg.name),
+            version: Cow::Borrowed(&pkg.version),
+            release: Cow::Borrowed(&pkg.release),
+            deps: Cow::Borrowed(&pkg.deps),
+            env: Cow::Borrowed(&pkg.env),
+            exposes: Cow::Borrowed(&pkg.exposes),
+            exports: Cow::Borrowed(&pkg.exports),
+            path: Cow::Borrowed(&pkg.path),
+            svc_path: Cow::Borrowed(&pkg.svc_path),
+            svc_config_path: Cow::Borrowed(&pkg.svc_config_path),
+            svc_data_path: Cow::Borrowed(&pkg.svc_data_path),
+            svc_files_path: Cow::Borrowed(&pkg.svc_files_path),
+            svc_static_path: Cow::Borrowed(&pkg.svc_static_path),
+            svc_var_path: Cow::Borrowed(&pkg.svc_var_path),
+            svc_pid_file: Cow::Borrowed(&pkg.svc_pid_file),
+            svc_run: Cow::Borrowed(&pkg.svc_run),
+            svc_user: Cow::Borrowed(&pkg.svc_user),
+            svc_group: Cow::Borrowed(&pkg.svc_group),
+        }
+    }
+}
+
+impl<'a> Serialize for Package<'a> {
+    fn serialize<S>(&self, serializer: S) -> result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(20))?;
```

Comment:
I'm guessing there's not any good way to derive this `20` or even catching that it needs updating if the `Package` struct changes.

Reply:
I don't _think_ so, but I'll review the Serde docs again to make sure.

Reply:
Yeah, I looked a bit and it wasn't terribly clear. It said that some serializers needed to know the size _a priori_, but what happened if it was not required and potentially differed from the number actually added seemed ambiguous.

My guess is that it's just a hint and if it's too low we'll just incur an additional alloc, but I think to really know we'd have to use my old master's advice: "use the source, Luke"

Reply:
Yeah, it sounds like "some serializers" need the hint, but others' don't. Apparently, the JSON serializer is one of the latter, as I can pass `None` as the hint, and all the tests still pass.

If we were just doing a straight serialization of the struct, then we could just use the number of fields in the struct (though really, we'd probably just derive the serialization at that point). However, I envision adding new serialized fields that may be derived from existing fields (e.g., renaming of some fields, while keeping the old one for backwards compatibility, etc.), which means the value won't be so easily computable, apart from a programmer counting.

These types are pretty strongly JSON focused now, seeing as how they're ultimately validated by JSON schema :smile:... we could just drop the size hint for these implementations. I know that we _do_ add the hint in our other `Serialize` implementations, though... probably a good practice to keep them, just for future-proofing's sake; I don't know what formats we may end up wanting to serialize to in the future.